Alumno: **Ginés González Guirado**

# Homework 3

3.2. Use Excel or a hand calculator to verify that the following sets of vectors form an orthogonal basis for their space. All angles are specified in degrees.

D=4 case:

In [1]:
import numpy as np

C0 = np.array([ 1.0, 1.0, 1.0, 1.0])
C1 = np.array([ 1.0, 0.0,-1.0, 0.0])
S1 = np.array([ 0.0, 1.0, 0.0,-1.0])
C2 = np.array([ 1.0,-1.0, 1.0,-1.0])

A = np.array([C0,
              C1,
              S1,
              C2])
A

array([[ 1.,  1.,  1.,  1.],
       [ 1.,  0., -1.,  0.],
       [ 0.,  1.,  0., -1.],
       [ 1., -1.,  1., -1.]])

In [2]:
det = np.linalg.det(A)

print(det)

-7.999999999999998


D=5 case:

In [3]:
import numpy as np

a1 = np.cos(72.0*np.pi/180.0)
b1 = np.cos(144.0*np.pi/180.0)
c1 = np.cos(216.0*np.pi/180.0)
d1 = np.cos(288.0*np.pi/180.0)
e1 = np.cos(432.0*np.pi/180.0)
f1 = np.cos(576.0*np.pi/180.0)
a2 = np.sin(72.0*np.pi/180.0)
b2 = np.sin(144.0*np.pi/180.0)
c2 = np.sin(216.0*np.pi/180.0)
d2 = np.sin(288.0*np.pi/180.0)
e2 = np.sin(432.0*np.pi/180.0)
f2 = np.sin(576.0*np.pi/180.0)

C0 = np.array([ 1.0, 1.0, 1.0, 1.0, 1.0])
C1 = np.array([ 1.0,  a1,  b1,  c1,  d1])
S1 = np.array([ 0.0,  a2,  b2,  c2,  d2])
C2 = np.array([ 1.0,  b1,  d1,  e1,  f1])
S2 = np.array([ 0.0,  b2,  d2,  e2,  f2])

A = np.array([C0,
              C1,
              S1,
              C2,
              S2])
A

array([[ 1.        ,  1.        ,  1.        ,  1.        ,  1.        ],
       [ 1.        ,  0.30901699, -0.80901699, -0.80901699,  0.30901699],
       [ 0.        ,  0.95105652,  0.58778525, -0.58778525, -0.95105652],
       [ 1.        , -0.80901699,  0.30901699,  0.30901699, -0.80901699],
       [ 0.        ,  0.58778525, -0.95105652,  0.95105652, -0.58778525]])

In [4]:
det = np.linalg.det(A)

print(det)

13.975424859373687


El determinante es distinto de 0, por lo que sólo el sistema tiene solución única y es la solución trivial, por tanto los vectores son linealmente independientes y forman una base.

3.4 (a) Write a MATLAB script to create the sampled cosine vectors $C_K$ and the sampled sine vectors $S_K$ defined in text equation $[3.26]$ for the case: $K =5$. In this formula, $K$ is called the "harmonic number". Accordingly, $S_2$ is the "second harmonic" of the sine function.

In [5]:
n=5    #Hasta donde llegan los vectors de base K=5
D=11   #Dimension tipo int
d=11.0  #Dimension como tipo float para poder dividir
pi = np.pi

#Genero los vectores qeu formarán la base

C0 = np.zeros(D)
C1 = np.zeros(D)
C2 = np.zeros(D)
C3 = np.zeros(D)
C4 = np.zeros(D)
C5 = np.zeros(D)

S1 = np.zeros(D)
S2 = np.zeros(D)
S3 = np.zeros(D)
S4 = np.zeros(D)
S5 = np.zeros(D)

#Bucle para calcular los vectores base

for i in range(0,D):
    k=0
    C0[i] = np.cos(2*i*pi*k/d)
    k=1
    C1[i] = np.cos(2*i*pi*k/d)
    S1[i] = np.sin(2*i*pi*k/d)
    k=2
    C2[i] = np.cos(2*i*pi*k/d)
    S2[i] = np.sin(2*i*pi*k/d)
    k=3
    C3[i] = np.cos(2*i*pi*k/d)
    S3[i] = np.sin(2*i*pi*k/d)
    k=4
    C4[i] = np.cos(2*i*pi*k/d)
    S4[i] = np.sin(2*i*pi*k/d)
    k=5
    C5[i] = np.cos(2*i*pi*k/d)
    S5[i] = np.sin(2*i*pi*k/d)
print("C_0 =",C0," C_1 =", C1," C_2 =", C2," C_3 =", C3," C_4 =", C4," C_5 =", C5," S_1 =", S1," S_2 =", S2," S_3 =", S3," S_4 =", S4," S_5 =", S5)

C_0 = [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]  C_1 = [ 1.          0.84125353  0.41541501 -0.14231484 -0.65486073 -0.95949297
 -0.95949297 -0.65486073 -0.14231484  0.41541501  0.84125353]  C_2 = [ 1.          0.41541501 -0.65486073 -0.95949297 -0.14231484  0.84125353
  0.84125353 -0.14231484 -0.95949297 -0.65486073  0.41541501]  C_3 = [ 1.         -0.14231484 -0.95949297  0.41541501  0.84125353 -0.65486073
 -0.65486073  0.84125353  0.41541501 -0.95949297 -0.14231484]  C_4 = [ 1.         -0.65486073 -0.14231484  0.84125353 -0.95949297  0.41541501
  0.41541501 -0.95949297  0.84125353 -0.14231484 -0.65486073]  C_5 = [ 1.         -0.95949297  0.84125353 -0.65486073  0.41541501 -0.14231484
 -0.14231484  0.41541501 -0.65486073  0.84125353 -0.95949297]  S_1 = [ 0.          0.54064082  0.909632    0.98982144  0.75574957  0.28173256
 -0.28173256 -0.75574957 -0.98982144 -0.909632   -0.54064082]  S_2 = [ 0.          0.909632    0.75574957 -0.28173256 -0.98982144 -0.54064082
  0.54064082  0.98982144  0

(b) Assemble the resulting column vectors into a matrix as indicated in text equation $[3.27]$.

Convierto los los arrays fila en columna de esta forma:

In [6]:
C00 = np.array([C0])  #Convierto array 1D en 2D
print(C0)
print(C00)
print(C0.shape)   #Forma del array
print(C00.shape)
print(C0.T)
print(C00.T)   #Traspuesta del array
print(C00.T.shape)   #Forma del array traspuesto (columna)

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]
(11,)
(1, 11)
[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
[[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]
(11, 1)


In [7]:
C00 = np.array([C0])
C000 = C00.T    #Array columna

C11 = np.array([C1])
C111 = C11.T

C22 = np.array([C2])
C222 = C22.T

C33 = np.array([C3])
C333 = C33.T

C44 = np.array([C4])
C444 = C44.T

C55 = np.array([C5])
C555 = C55.T

S11 = np.array([S1])
S111 = S11.T

S22 = np.array([S2])
S222 = S22.T

S33 = np.array([S3])
S333 = S33.T

S44 = np.array([S4])
S444 = S44.T

S55 = np.array([S5])
S555 = S55.T

In [8]:
A = np.hstack([C000, C111, S111, C222, S222, C333, S333, C444, S444, C555, S555])  #Uno arrays columna en matriz
print(A)
print(A.shape)

[[ 1.          1.          0.          1.          0.          1.
   0.          1.          0.          1.          0.        ]
 [ 1.          0.84125353  0.54064082  0.41541501  0.909632   -0.14231484
   0.98982144 -0.65486073  0.75574957 -0.95949297  0.28173256]
 [ 1.          0.41541501  0.909632   -0.65486073  0.75574957 -0.95949297
  -0.28173256 -0.14231484 -0.98982144  0.84125353 -0.54064082]
 [ 1.         -0.14231484  0.98982144 -0.95949297 -0.28173256  0.41541501
  -0.909632    0.84125353  0.54064082 -0.65486073  0.75574957]
 [ 1.         -0.65486073  0.75574957 -0.14231484 -0.98982144  0.84125353
   0.54064082 -0.95949297  0.28173256  0.41541501 -0.909632  ]
 [ 1.         -0.95949297  0.28173256  0.84125353 -0.54064082 -0.65486073
   0.75574957  0.41541501 -0.909632   -0.14231484  0.98982144]
 [ 1.         -0.95949297 -0.28173256  0.84125353  0.54064082 -0.65486073
  -0.75574957  0.41541501  0.909632   -0.14231484 -0.98982144]
 [ 1.         -0.65486073 -0.75574957 -0.14231484

(c) Use matrix mulitplication to find the length of each of the column vectors in the matrix created in part (b).

In [9]:
# Multiplico la traspuesta de A por A, de forma que se obtiene el modulo al cuadrado 
# o la longitud al cuadrado de cada vector en los elementos de la diagonal
AI = A.T @ A
AI

array([[ 1.10000000e+01, -2.22044605e-16,  5.55111512e-16,
         5.55111512e-17,  9.99200722e-16, -1.74860126e-15,
         1.66533454e-15,  3.33066907e-15, -2.22044605e-16,
         4.44089210e-16, -7.77156117e-16],
       [-2.22044605e-16,  5.50000000e+00,  4.63954990e-16,
        -3.39120447e-16,  4.08082137e-16, -3.02208085e-15,
        -4.27413454e-16,  1.19246744e-16, -3.78176620e-16,
         2.19318353e-15, -1.20718207e-15],
       [ 5.55111512e-16,  4.63954990e-16,  5.50000000e+00,
         1.17796665e-16,  4.06907775e-17,  1.15311866e-15,
        -6.47524763e-16, -1.26669862e-15, -5.76617790e-16,
        -1.55100009e-16,  2.71603010e-15],
       [ 5.55111512e-17, -3.39120447e-16,  1.17796665e-16,
         5.50000000e+00, -3.94310884e-17, -1.16381130e-15,
         1.36353558e-15, -2.29579339e-16, -9.75855704e-16,
        -5.57589511e-15, -3.87612643e-16],
       [ 9.99200722e-16,  4.08082137e-16,  4.06907775e-17,
        -3.94310884e-17,  5.50000000e+00,  2.06524409e-15,
  

Ahora, selecciono los elementos que son el módulo al cuadrado o longitud al cuadrado de cada uno de los vectores de la base, es decir, los de la diagonal principal.

In [10]:
LLC0 = AI[0,0]
LLC1 = AI[1,1]
LLS1 = AI[2,2]
LLC2 = AI[3,3]
LLS2 = AI[4,4]
LLC3 = AI[5,5]
LLS3 = AI[6,6]
LLC4 = AI[7,7]
LLS4 = AI[8,8]
LLC5 = AI[9,9]
LLS5 = AI[10,10]

Haciendo la raíz cuadrada obtenemos los módulos de cada uno de los vectores de la base.

In [11]:
LC0 = np.sqrt(LLC0)
LC1 = np.sqrt(LLC1)
LS1 = np.sqrt(LLS1)
LC2 = np.sqrt(LLC2)
LS2 = np.sqrt(LLS2)
LC3 = np.sqrt(LLC3)
LS3 = np.sqrt(LLS3)
LC4 = np.sqrt(LLC4)
LS4 = np.sqrt(LLS4)
LC5 = np.sqrt(LLC5)
LS5 = np.sqrt(LLS5)

print("Longitud C0 =",LC0)
print("Longitud C1 =",LC1)
print("Longitud S1 =",LS1)
print("Longitud C2 =",LC2)
print("Longitud S2 =",LS2)
print("Longitud C3 =",LC3)
print("Longitud S3 =",LS3)
print("Longitud C4 =",LC4)
print("Longitud S4 =",LS4)
print("Longitud C5 =",LC5)
print("Longitud S5 =",LS5)

Longitud C0 = 3.3166247903554
Longitud C1 = 2.345207879911715
Longitud S1 = 2.345207879911715
Longitud C2 = 2.3452078799117153
Longitud S2 = 2.3452078799117144
Longitud C3 = 2.3452078799117153
Longitud S3 = 2.3452078799117144
Longitud C4 = 2.345207879911715
Longitud S4 = 2.345207879911715
Longitud C5 = 2.345207879911715
Longitud S5 = 2.3452078799117144


(d) Use your script to check your results from problem 3.2 above.

Caso D=4:

In [12]:
D=4
d=4.0
pi = np.pi

C0 = np.zeros(D)
C1 = np.zeros(D)
C2 = np.zeros(D)

S1 = np.zeros(D)

for i in range(0,D):
    k=0
    C0[i] = np.cos(2*i*pi*k/d)
    k=1
    C1[i] = np.cos(2*i*pi*k/d)
    S1[i] = np.sin(2*i*pi*k/d)
    k=2
    C2[i] = np.cos(2*i*pi*k/d)

print("C_0 =",C0," C_1 =", C1," C_2 =", C2,"S_1 =", S1)

C_0 = [1. 1. 1. 1.]  C_1 = [ 1.0000000e+00  6.1232340e-17 -1.0000000e+00 -1.8369702e-16]  C_2 = [ 1. -1.  1. -1.] S_1 = [ 0.0000000e+00  1.0000000e+00  1.2246468e-16 -1.0000000e+00]


Los vectores forman una base si son linealmente independientes, es decir, el siguiente determinante tiene que ser distinto de cero. Entonces, tendríamos solución única, la cuál sería la trivial y los vectores serían linealmente independientes, formando así una base. 

Convierto los arrays fila en columna y genero la matriz de la que calcularé el determinante para comprobar que es distinto de 0.

In [13]:
C00 = np.array([C0])
C000 = C00.T

C11 = np.array([C1])
C111 = C11.T

C22 = np.array([C2])
C222 = C22.T

S11 = np.array([S1])
S111 = S11.T

B = np.hstack([C000, C111, S111, C222])
print(B)
print(B.shape)

[[ 1.0000000e+00  1.0000000e+00  0.0000000e+00  1.0000000e+00]
 [ 1.0000000e+00  6.1232340e-17  1.0000000e+00 -1.0000000e+00]
 [ 1.0000000e+00 -1.0000000e+00  1.2246468e-16  1.0000000e+00]
 [ 1.0000000e+00 -1.8369702e-16 -1.0000000e+00 -1.0000000e+00]]
(4, 4)


In [14]:
det = np.linalg.det(B)

print(det)

-7.999999999999998


Compruebo que son ortogonales, haciendo el producto escalar de cada uno de ellos por los demás.

In [15]:
print("C0.C1 =",C0 @ C1)
print("C0.C2 =",C0 @ C2)
print("C0.S1 =",C0 @ S1)
print("C1.C2 =",C1 @ C2)
print("C1.S1 =",C1 @ S1)
print("C2.S1 =",C2 @ S1)

C0.C1 = -1.224646799147353e-16
C0.C2 = 0.0
C0.S1 = 1.2246467991473532e-16
C1.C2 = 1.224646799147353e-16
C1.S1 = 1.2246467991473532e-16
C2.S1 = 1.2246467991473532e-16


Caso D=5:

In [16]:
D=5
d=5.0
pi = np.pi

C0 = np.zeros(D)
C1 = np.zeros(D)
C2 = np.zeros(D)

S1 = np.zeros(D)
S2 = np.zeros(D)

for i in range(0,D):
    k=0
    C0[i] = np.cos(2*i*pi*k/d)
    k=1
    C1[i] = np.cos(2*i*pi*k/d)
    S1[i] = np.sin(2*i*pi*k/d)
    k=2
    C2[i] = np.cos(2*i*pi*k/d)
    S2[i] = np.sin(2*i*pi*k/d)

print("C_0 =",C0," C_1 =", C1," C_2 =", C2," S_1 =", S1," S_2 =", S2)

C_0 = [1. 1. 1. 1. 1.]  C_1 = [ 1.          0.30901699 -0.80901699 -0.80901699  0.30901699]  C_2 = [ 1.         -0.80901699  0.30901699  0.30901699 -0.80901699]  S_1 = [ 0.          0.95105652  0.58778525 -0.58778525 -0.95105652]  S_2 = [ 0.          0.58778525 -0.95105652  0.95105652 -0.58778525]


Los vectores forman una base si son linealmente independientes, es decir, el siguiente determinante tiene que ser distinto de cero. Entonces, tendríamos solución única, la cuál sería la trivial y los vectores serían linealmente independientes, formando así una base. 

Convierto los arrays fila en columna y genero la matriz de la que calcularé el determinante para comprobar que es distinto de 0.

In [17]:
C00 = np.array([C0])
C000 = C00.T

C11 = np.array([C1])
C111 = C11.T

C22 = np.array([C2])
C222 = C22.T

S11 = np.array([S1])
S111 = S11.T

S22 = np.array([S2])
S222 = S22.T

B = np.hstack([C000, C111, S111, C222, S222])
print(B)
print(B.shape)

[[ 1.          1.          0.          1.          0.        ]
 [ 1.          0.30901699  0.95105652 -0.80901699  0.58778525]
 [ 1.         -0.80901699  0.58778525  0.30901699 -0.95105652]
 [ 1.         -0.80901699 -0.58778525  0.30901699  0.95105652]
 [ 1.          0.30901699 -0.95105652 -0.80901699 -0.58778525]]
(5, 5)


In [18]:
det = np.linalg.det(B)

print(det)

13.975424859373682


Compruebo que son ortogonales, haciendo el producto escalar de cada uno de ellos por los demás.

In [19]:
print("C0.C1 =",C0 @ C1)
print("C0.C2 =",C0 @ C2)
print("C0.S1 =",C0 @ S1)
print("C0.S2 =",C0 @ S2)
print("C1.C2 =",C1 @ C2)
print("C1.S1 =",C1 @ S1)
print("C1.S2 =",C1 @ S2)
print("C2.S1 =",C2 @ S1)
print("C2.S2 =",C2 @ S2)
print("C1.S2 =",C1 @ S2)

C0.C1 = -1.1102230246251565e-16
C0.C2 = -5.551115123125783e-17
C0.S1 = 1.1102230246251565e-16
C0.S2 = 2.220446049250313e-16
C1.C2 = 1.1102230246251565e-16
C1.S1 = 0.0
C1.S2 = 1.1102230246251565e-16
C2.S1 = 1.1102230246251565e-16
C2.S2 = 2.7755575615628914e-16
C1.S2 = 1.1102230246251565e-16


3.5. Write a MATLAB procedure to compute the ortho-normal trigonometrical matrix defined in text eqn. 3.39 for any vector of sample times. Verify your result for the case of D=2,3,4,5.

Calculo los vectores normalizados.

In [20]:
# Caso D=11, si D es mayor solamente hay que añadir mas vectores
n=5
D=11
d=11.0
pi = np.pi

c0 = np.zeros(D)
c1 = np.zeros(D)
c2 = np.zeros(D)
c3 = np.zeros(D)
c4 = np.zeros(D)
c5 = np.zeros(D)

s1 = np.zeros(D)
s2 = np.zeros(D)
s3 = np.zeros(D)
s4 = np.zeros(D)
s5 = np.zeros(D)

for i in range(0,D):
    k=0
    c0[i] = (1.0/np.sqrt(d))*np.cos(2*i*pi*k/d)
    k=1
    c1[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s1[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=2
    c2[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s2[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=3
    c3[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s3[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=4
    c4[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s4[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=5
    c5[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s5[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
print("c_0 =",c0," c_1 =", c1," c_2 =", c2," c_3 =", c3," c_4 =", c4," c_5 =", c5," s_1 =", s1," s_2 =", s2," s_3 =", s3," s_4 =", s4," s_5 =", s5)

c_0 = [0.30151134 0.30151134 0.30151134 0.30151134 0.30151134 0.30151134
 0.30151134 0.30151134 0.30151134 0.30151134 0.30151134]  c_1 = [ 0.42640143  0.35871171  0.17713356 -0.06068325 -0.27923356 -0.40912918
 -0.40912918 -0.27923356 -0.06068325  0.17713356  0.35871171]  c_2 = [ 0.42640143  0.17713356 -0.27923356 -0.40912918 -0.06068325  0.35871171
  0.35871171 -0.06068325 -0.40912918 -0.27923356  0.17713356]  c_3 = [ 0.42640143 -0.06068325 -0.40912918  0.17713356  0.35871171 -0.27923356
 -0.27923356  0.35871171  0.17713356 -0.40912918 -0.06068325]  c_4 = [ 0.42640143 -0.27923356 -0.06068325  0.35871171 -0.40912918  0.17713356
  0.17713356 -0.40912918  0.35871171 -0.06068325 -0.27923356]  c_5 = [ 0.42640143 -0.40912918  0.35871171 -0.27923356  0.17713356 -0.06068325
 -0.06068325  0.17713356 -0.27923356  0.35871171 -0.40912918]  s_1 = [ 0.          0.23053002  0.38786839  0.42206128  0.3222527   0.12013117
 -0.12013117 -0.3222527  -0.42206128 -0.38786839 -0.23053002]  s_2 = [ 0.       

In [21]:
A = np.array([c0,
              c1,
              s1,
              c2,
              s2,
              c3,
              s3,
              c4,
              s4,
              c5,
              s5])  #Matriz con los vectores base en fila
print(A)
print(A.shape)

[[ 0.30151134  0.30151134  0.30151134  0.30151134  0.30151134  0.30151134
   0.30151134  0.30151134  0.30151134  0.30151134  0.30151134]
 [ 0.42640143  0.35871171  0.17713356 -0.06068325 -0.27923356 -0.40912918
  -0.40912918 -0.27923356 -0.06068325  0.17713356  0.35871171]
 [ 0.          0.23053002  0.38786839  0.42206128  0.3222527   0.12013117
  -0.12013117 -0.3222527  -0.42206128 -0.38786839 -0.23053002]
 [ 0.42640143  0.17713356 -0.27923356 -0.40912918 -0.06068325  0.35871171
   0.35871171 -0.06068325 -0.40912918 -0.27923356  0.17713356]
 [ 0.          0.38786839  0.3222527  -0.12013117 -0.42206128 -0.23053002
   0.23053002  0.42206128  0.12013117 -0.3222527  -0.38786839]
 [ 0.42640143 -0.06068325 -0.40912918  0.17713356  0.35871171 -0.27923356
  -0.27923356  0.35871171  0.17713356 -0.40912918 -0.06068325]
 [ 0.          0.42206128 -0.12013117 -0.38786839  0.23053002  0.3222527
  -0.3222527  -0.23053002  0.38786839  0.12013117 -0.42206128]
 [ 0.42640143 -0.27923356 -0.06068325  0.3

In [22]:
# Genero un vector (Que en un ejercicio se tendría que dar)
v = np.linspace(0,1,D)
v0 = np.array([v])
v00 = v0.T

#Forward DFT
f = np.sqrt(2.0/d)*A @ v00
f

array([[ 0.70710678],
       [-0.1       ],
       [-0.34056872],
       [-0.1       ],
       [-0.15560304],
       [-0.1       ],
       [-0.08665049],
       [-0.1       ],
       [-0.04566847],
       [-0.1       ],
       [-0.01437783]])

Caso D=2:

In [23]:
n=1
D=2
d=2.0
pi = np.pi

c0 = np.zeros(D)
c1 = np.zeros(D)

for i in range(0,D):
    k=0
    c0[i] = (1.0/np.sqrt(d))*np.cos(2*i*pi*k/d)
    k=1
    c1[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)

print("c_0 =",c0," c_1 =", c1)

c_0 = [0.70710678 0.70710678]  c_1 = [ 1. -1.]


In [24]:
A = np.array([c0,c1])

print(A)
print(A.shape)

[[ 0.70710678  0.70710678]
 [ 1.         -1.        ]]
(2, 2)


In [25]:
v = np.linspace(0.1,1,D)
v0 = np.array([v])
v00 = v0.T

f = np.sqrt(2.0/d)*A @ v00
f

array([[ 0.77781746],
       [-0.9       ]])

Caso D=3:

In [26]:
n=1
D=3
d=3.0
pi = np.pi

c0 = np.zeros(D)
c1 = np.zeros(D)
s1 = np.zeros(D)

for i in range(0,D):
    k=0
    c0[i] = (1.0/np.sqrt(d))*np.cos(2*i*pi*k/d)
    k=1
    c1[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s1[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)

print("c_0 =",c0," c_1 =", c1," s_1 =",s1)

c_0 = [0.57735027 0.57735027 0.57735027]  c_1 = [ 0.81649658 -0.40824829 -0.40824829]  s_1 = [ 0.          0.70710678 -0.70710678]


In [27]:
A = np.array([c0,c1,s1])

print(A)
print(A.shape)

[[ 0.57735027  0.57735027  0.57735027]
 [ 0.81649658 -0.40824829 -0.40824829]
 [ 0.          0.70710678 -0.70710678]]
(3, 3)


In [28]:
v = np.linspace(0.1,1,D)
v0 = np.array([v])
v00 = v0.T

f = np.sqrt(2.0/d)*A @ v00
f

array([[ 0.77781746],
       [-0.45      ],
       [-0.25980762]])

Caso D=4:

In [29]:
n=2
D=4
d=4.0
pi = np.pi

c0 = np.zeros(D)
c1 = np.zeros(D)
c2 = np.zeros(D)

s1 = np.zeros(D)

for i in range(0,D):
    k=0
    c0[i] = (1.0/np.sqrt(d))*np.cos(2*i*pi*k/d)
    k=1
    c1[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s1[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=2
    c2[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)

print("c_0 =",c0," c_1 =", c1," c_2 =", c2," s_1 =", s1)

c_0 = [0.5 0.5 0.5 0.5]  c_1 = [ 7.07106781e-01  4.32978028e-17 -7.07106781e-01 -1.29893408e-16]  c_2 = [ 0.70710678 -0.70710678  0.70710678 -0.70710678]  s_1 = [ 0.00000000e+00  7.07106781e-01  8.65956056e-17 -7.07106781e-01]


In [30]:
A = np.array([c0,c1,s1,c2])

print(A)
print(A.shape)

[[ 5.00000000e-01  5.00000000e-01  5.00000000e-01  5.00000000e-01]
 [ 7.07106781e-01  4.32978028e-17 -7.07106781e-01 -1.29893408e-16]
 [ 0.00000000e+00  7.07106781e-01  8.65956056e-17 -7.07106781e-01]
 [ 7.07106781e-01 -7.07106781e-01  7.07106781e-01 -7.07106781e-01]]
(4, 4)


In [31]:
v = np.linspace(0.1,1,D)
v0 = np.array([v])
v00 = v0.T

f = np.sqrt(2.0/d)*A @ v00
f

array([[ 0.77781746],
       [-0.3       ],
       [-0.3       ],
       [-0.3       ]])

Caso D=5:

In [32]:
n=2
D=5
d=5.0
pi = np.pi

c0 = np.zeros(D)
c1 = np.zeros(D)
c2 = np.zeros(D)

s1 = np.zeros(D)
s2 = np.zeros(D)

for i in range(0,D):
    k=0
    c0[i] = (1.0/np.sqrt(d))*np.cos(2*i*pi*k/d)
    k=1
    c1[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s1[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=2
    c2[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s2[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)

print("c_0 =",c0," c_1 =", c1," c_2 =", c2," s_1 =", s1," s_2 =", s2)

c_0 = [0.4472136 0.4472136 0.4472136 0.4472136 0.4472136]  c_1 = [ 0.63245553  0.19543951 -0.51166727 -0.51166727  0.19543951]  c_2 = [ 0.63245553 -0.51166727  0.19543951  0.19543951 -0.51166727]  s_1 = [ 0.          0.60150096  0.37174803 -0.37174803 -0.60150096]  s_2 = [ 0.          0.37174803 -0.60150096  0.60150096 -0.37174803]


In [33]:
A = np.array([c0,c1,s1,c2,s2])

print(A)
print(A.shape)

[[ 0.4472136   0.4472136   0.4472136   0.4472136   0.4472136 ]
 [ 0.63245553  0.19543951 -0.51166727 -0.51166727  0.19543951]
 [ 0.          0.60150096  0.37174803 -0.37174803 -0.60150096]
 [ 0.63245553 -0.51166727  0.19543951  0.19543951 -0.51166727]
 [ 0.          0.37174803 -0.60150096  0.60150096 -0.37174803]]
(5, 5)


In [34]:
v = np.linspace(0.1,1,D)
v0 = np.array([v])
v00 = v0.T

f = np.sqrt(2.0/d)*A @ v00
f

array([[ 0.77781746],
       [-0.225     ],
       [-0.30968593],
       [-0.225     ],
       [-0.07310693]])

3.6 Write a MATLAB procedure to implement the forward DFT (text eqn. 3.39) and inverse DFT (text eqn. 3.38) for any data vector of length D. Verify the routine for the following 2 data vectors: $v1 = [3, -2, 6]$, $v2 = [-5, 0, 3, 8]$

A procedure to implement the forward DFT is:

In [35]:
# Primero definimos los vectores de la base, por ejemplo, para D=11
n=5
D=11
d=11.0
pi = np.pi

c0 = np.zeros(D)
c1 = np.zeros(D)
c2 = np.zeros(D)
c3 = np.zeros(D)
c4 = np.zeros(D)
c5 = np.zeros(D)

s1 = np.zeros(D)
s2 = np.zeros(D)
s3 = np.zeros(D)
s4 = np.zeros(D)
s5 = np.zeros(D)

for i in range(0,D):
    k=0
    c0[i] = (1.0/np.sqrt(d))*np.cos(2*i*pi*k/d)
    k=1
    c1[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s1[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=2
    c2[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s2[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=3
    c3[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s3[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=4
    c4[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s4[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=5
    c5[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s5[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
print("c_0 =",c0," c_1 =", c1," c_2 =", c2," c_3 =", c3," c_4 =", c4," c_5 =", c5," s_1 =", s1," s_2 =", s2," s_3 =", s3," s_4 =", s4," s_5 =", s5)

c_0 = [0.30151134 0.30151134 0.30151134 0.30151134 0.30151134 0.30151134
 0.30151134 0.30151134 0.30151134 0.30151134 0.30151134]  c_1 = [ 0.42640143  0.35871171  0.17713356 -0.06068325 -0.27923356 -0.40912918
 -0.40912918 -0.27923356 -0.06068325  0.17713356  0.35871171]  c_2 = [ 0.42640143  0.17713356 -0.27923356 -0.40912918 -0.06068325  0.35871171
  0.35871171 -0.06068325 -0.40912918 -0.27923356  0.17713356]  c_3 = [ 0.42640143 -0.06068325 -0.40912918  0.17713356  0.35871171 -0.27923356
 -0.27923356  0.35871171  0.17713356 -0.40912918 -0.06068325]  c_4 = [ 0.42640143 -0.27923356 -0.06068325  0.35871171 -0.40912918  0.17713356
  0.17713356 -0.40912918  0.35871171 -0.06068325 -0.27923356]  c_5 = [ 0.42640143 -0.40912918  0.35871171 -0.27923356  0.17713356 -0.06068325
 -0.06068325  0.17713356 -0.27923356  0.35871171 -0.40912918]  s_1 = [ 0.          0.23053002  0.38786839  0.42206128  0.3222527   0.12013117
 -0.12013117 -0.3222527  -0.42206128 -0.38786839 -0.23053002]  s_2 = [ 0.       

In [36]:
A = np.array([c0,c1,s1,c2,s2,c3,s3,c4,s4,c5,s5])

print(A)
print(A.shape)

[[ 0.30151134  0.30151134  0.30151134  0.30151134  0.30151134  0.30151134
   0.30151134  0.30151134  0.30151134  0.30151134  0.30151134]
 [ 0.42640143  0.35871171  0.17713356 -0.06068325 -0.27923356 -0.40912918
  -0.40912918 -0.27923356 -0.06068325  0.17713356  0.35871171]
 [ 0.          0.23053002  0.38786839  0.42206128  0.3222527   0.12013117
  -0.12013117 -0.3222527  -0.42206128 -0.38786839 -0.23053002]
 [ 0.42640143  0.17713356 -0.27923356 -0.40912918 -0.06068325  0.35871171
   0.35871171 -0.06068325 -0.40912918 -0.27923356  0.17713356]
 [ 0.          0.38786839  0.3222527  -0.12013117 -0.42206128 -0.23053002
   0.23053002  0.42206128  0.12013117 -0.3222527  -0.38786839]
 [ 0.42640143 -0.06068325 -0.40912918  0.17713356  0.35871171 -0.27923356
  -0.27923356  0.35871171  0.17713356 -0.40912918 -0.06068325]
 [ 0.          0.42206128 -0.12013117 -0.38786839  0.23053002  0.3222527
  -0.3222527  -0.23053002  0.38786839  0.12013117 -0.42206128]
 [ 0.42640143 -0.27923356 -0.06068325  0.3

In [37]:
# Genero un vector (Que en un ejercicio se tendría que dar)
v = np.linspace(0,1,D)
v0 = np.array([v])
v00 = v0.T

# Forward DFT
f = np.sqrt(2.0/d)*A @ v00
f

array([[ 0.70710678],
       [-0.1       ],
       [-0.34056872],
       [-0.1       ],
       [-0.15560304],
       [-0.1       ],
       [-0.08665049],
       [-0.1       ],
       [-0.04566847],
       [-0.1       ],
       [-0.01437783]])

Inverse DFT:

In [38]:
# Primero definimos los vectores de la base, por ejemplo, para D=11
n=5
D=11
d=11.0
pi = np.pi

c0 = np.zeros(D)
c1 = np.zeros(D)
c2 = np.zeros(D)
c3 = np.zeros(D)
c4 = np.zeros(D)
c5 = np.zeros(D)

s1 = np.zeros(D)
s2 = np.zeros(D)
s3 = np.zeros(D)
s4 = np.zeros(D)
s5 = np.zeros(D)

for i in range(0,D):
    k=0
    c0[i] = (1.0/np.sqrt(d))*np.cos(2*i*pi*k/d)
    k=1
    c1[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s1[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=2
    c2[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s2[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=3
    c3[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s3[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=4
    c4[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s4[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=5
    c5[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s5[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
print("c_0 =",c0," c_1 =", c1," c_2 =", c2," c_3 =", c3," c_4 =", c4," c_5 =", c5," s_1 =", s1," s_2 =", s2," s_3 =", s3," s_4 =", s4," s_5 =", s5)

c_0 = [0.30151134 0.30151134 0.30151134 0.30151134 0.30151134 0.30151134
 0.30151134 0.30151134 0.30151134 0.30151134 0.30151134]  c_1 = [ 0.42640143  0.35871171  0.17713356 -0.06068325 -0.27923356 -0.40912918
 -0.40912918 -0.27923356 -0.06068325  0.17713356  0.35871171]  c_2 = [ 0.42640143  0.17713356 -0.27923356 -0.40912918 -0.06068325  0.35871171
  0.35871171 -0.06068325 -0.40912918 -0.27923356  0.17713356]  c_3 = [ 0.42640143 -0.06068325 -0.40912918  0.17713356  0.35871171 -0.27923356
 -0.27923356  0.35871171  0.17713356 -0.40912918 -0.06068325]  c_4 = [ 0.42640143 -0.27923356 -0.06068325  0.35871171 -0.40912918  0.17713356
  0.17713356 -0.40912918  0.35871171 -0.06068325 -0.27923356]  c_5 = [ 0.42640143 -0.40912918  0.35871171 -0.27923356  0.17713356 -0.06068325
 -0.06068325  0.17713356 -0.27923356  0.35871171 -0.40912918]  s_1 = [ 0.          0.23053002  0.38786839  0.42206128  0.3222527   0.12013117
 -0.12013117 -0.3222527  -0.42206128 -0.38786839 -0.23053002]  s_2 = [ 0.       

Convierto los los arrays fila en columna:

In [39]:
c00 = np.array([c0])
c000 = c00.T

c11 = np.array([c1])
c111 = c11.T

c22 = np.array([c2])
c222 = c22.T

c33 = np.array([c3])
c333 = c33.T

c44 = np.array([c4])
c444 = c44.T

c55 = np.array([c5])
c555 = c55.T

s11 = np.array([s1])
s111 = s11.T

s22 = np.array([s2])
s222 = s22.T

s33 = np.array([s3])
s333 = s33.T

s44 = np.array([s4])
s444 = s44.T

s55 = np.array([s5])
s555 = s55.T

# Genero la matriz con vectores base columna
A = np.hstack([c000, c111, s111, c222, s222, c333, s333, c444, s444, c555, s555])
print(A)
print(A.shape)

[[ 0.30151134  0.42640143  0.          0.42640143  0.          0.42640143
   0.          0.42640143  0.          0.42640143  0.        ]
 [ 0.30151134  0.35871171  0.23053002  0.17713356  0.38786839 -0.06068325
   0.42206128 -0.27923356  0.3222527  -0.40912918  0.12013117]
 [ 0.30151134  0.17713356  0.38786839 -0.27923356  0.3222527  -0.40912918
  -0.12013117 -0.06068325 -0.42206128  0.35871171 -0.23053002]
 [ 0.30151134 -0.06068325  0.42206128 -0.40912918 -0.12013117  0.17713356
  -0.38786839  0.35871171  0.23053002 -0.27923356  0.3222527 ]
 [ 0.30151134 -0.27923356  0.3222527  -0.06068325 -0.42206128  0.35871171
   0.23053002 -0.40912918  0.12013117  0.17713356 -0.38786839]
 [ 0.30151134 -0.40912918  0.12013117  0.35871171 -0.23053002 -0.27923356
   0.3222527   0.17713356 -0.38786839 -0.06068325  0.42206128]
 [ 0.30151134 -0.40912918 -0.12013117  0.35871171  0.23053002 -0.27923356
  -0.3222527   0.17713356  0.38786839 -0.06068325 -0.42206128]
 [ 0.30151134 -0.27923356 -0.3222527  -0.

In [40]:
# Genero un vector de los coeficientes (Que en un ejercicio se tendría que dar)
f = np.linspace(0.2,1,D)
f0 = np.array([f])
f00 = f0.T

# Inverse DFT
v = np.sqrt(d/2.0)*A @ f00
v

array([[ 3.14142136],
       [ 1.35121847],
       [-0.78351464],
       [ 0.766699  ],
       [-0.4948171 ],
       [ 0.59471497],
       [-0.52514658],
       [ 0.60113929],
       [-0.83564678],
       [ 0.9029087 ],
       [-3.16334176]])

Compruebo para $v_1 = [3,-2,6]$:

In [41]:
# Hago forward DFT

n=1
D=3
d=3.0
pi = np.pi

c0 = np.zeros(D)
c1 = np.zeros(D)

s1 = np.zeros(D)

for i in range(0,D):
    k=0
    c0[i] = (1.0/np.sqrt(d))*np.cos(2*i*pi*k/d)
    k=1
    c1[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s1[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)

print("c_0 =",c0," c_1 =", c1," s_1 =", s1)

c_0 = [0.57735027 0.57735027 0.57735027]  c_1 = [ 0.81649658 -0.40824829 -0.40824829]  s_1 = [ 0.          0.70710678 -0.70710678]


In [42]:
A = np.array([c0,c1,s1])

print(A)
print(A.shape)

[[ 0.57735027  0.57735027  0.57735027]
 [ 0.81649658 -0.40824829 -0.40824829]
 [ 0.          0.70710678 -0.70710678]]
(3, 3)


In [43]:
v = np.array([3,-2,6])
v0 = np.array([v])
v00 = v0.T

# Forward DFT
f = np.sqrt(2.0/d)*A @ v00
f

array([[ 3.29983165],
       [ 0.66666667],
       [-4.61880215]])

Compruebo para $v_2 = [-5,0,3,8]$:

In [44]:
n=2
D=4
d=4.0
pi = np.pi

c0 = np.zeros(D)
c1 = np.zeros(D)
c2 = np.zeros(D)

s1 = np.zeros(D)

for i in range(0,D):
    k=0
    c0[i] = (1.0/np.sqrt(d))*np.cos(2*i*pi*k/d)
    k=1
    c1[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)
    s1[i] = (1.0/np.sqrt(d/2.0))*np.sin(2*i*pi*k/d)
    k=2
    c2[i] = (1.0/np.sqrt(d/2.0))*np.cos(2*i*pi*k/d)

print("c_0 =",c0," c_1 =", c1," c_2 =", c2," s_1 =", s1)

c_0 = [0.5 0.5 0.5 0.5]  c_1 = [ 7.07106781e-01  4.32978028e-17 -7.07106781e-01 -1.29893408e-16]  c_2 = [ 0.70710678 -0.70710678  0.70710678 -0.70710678]  s_1 = [ 0.00000000e+00  7.07106781e-01  8.65956056e-17 -7.07106781e-01]


In [45]:
A = np.array([c0,c1,s1,c2])

print(A)
print(A.shape)

[[ 5.00000000e-01  5.00000000e-01  5.00000000e-01  5.00000000e-01]
 [ 7.07106781e-01  4.32978028e-17 -7.07106781e-01 -1.29893408e-16]
 [ 0.00000000e+00  7.07106781e-01  8.65956056e-17 -7.07106781e-01]
 [ 7.07106781e-01 -7.07106781e-01  7.07106781e-01 -7.07106781e-01]]
(4, 4)


In [46]:
v = np.array([-5,0,3,8])
v0 = np.array([v])
v00 = v0.T

# Forward DFT
f = np.sqrt(2.0/d)*A @ v00
f

array([[ 2.12132034],
       [-4.        ],
       [-4.        ],
       [-5.        ]])